Tone down offstrategy log message

bhalevy opened the pull request:
In many cases we trigger offstrategy compaction opportunistically
also when there's nothing to do.  In this case we still print
to the log lots of info-level message and call
`run_offstrategy_compaction` that wastes more cpu cycles
on learning that it has nothing to do.

This change bails out early if the maintenance set is empty
and prints a "Skipping off-strategy compaction" message in debug
level instead.

Fixes #13466

Also, add an group_id class and return it from compaction_group and table_state.
Use that to identify the compaction_group / table_state by "ks_name.cf_name compaction_group=idx/total" in log messages.

Fixes #13467

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -1265,6 +1265,13 @@ class offstrategy_compaction_task_executor : public compaction_task_executor {
                 table_state& t = *_compacting_table;
                 auto maintenance_sstables = t.maintenance_sstable_set().all();
+                auto size = maintenance_sstables->size();
+                if (!size) {
```
we added a sstable_set::size() method, let's take this opportunity to kill

auto maintenance_sstables = t.maintenance_sstable_set().all();

and retrieve size directly from set, with  t.maintenance_sstable_set().size();

bhalevy replied:
fixed in v2

raphaelsc commented on replica/table.cc:
```diff
@@ -1274,6 +1274,11 @@ future<bool> table::perform_offstrategy_compaction() {
+            tlogger.debug("Skipping off-strategy compaction for {}.{}, No candidates were found",
+                    schema()->ks_name(), schema()->cf_name());
+            co_return;
+        }
```
is this left over? once we agreed this logic should only exist in compaction_manager, so as not to have knowledge dupĺicated about what triggers off-strategy. we had a check like this in replica::table, but we removed it.

bhalevy replied:
I'll keep it in the compaction_manager layer in next version

bhalevy replied:
The reason it's more efficient to skip the function at the higher layers is that we actually do quite a lot, like creating the task and then acquire the semaphore just to find out we have nothing to do.

raphaelsc replied:
I have an idea: add `bool needs_offstrategy_compaction()` to table_state interface, which can be used by both compaction manager and replica::table. so no knowledge duplication.

raphaelsc replied:
I looked at offstrategy_compaction_task_executor::do_run() and what we can possibly do is to move the (!size) check to the beginning of the function, meaning it won't have to unnecessarily acquire the semaphore.

bhalevy replied:
yeah, let's consider this as a follow up.
I agree it's not on any hot path so maybe not worth the effort.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -1265,6 +1265,13 @@ class offstrategy_compaction_task_executor : public compaction_task_executor {
+                if (!size) {
+                    cmlog.debug("Skipping off-strategy compaction for {}.{}, {} candidates were found",
+                            t.schema()->ks_name(), t.schema()->cf_name(), size);
+                    finish_compaction();
```
Sorry, this was left over from my first take on it.
Moved skipping to the table layer so we can leave this code as it is

bhalevy replied:
Kept here in v2, saying "No candidates were found"

raphaelsc commented on replica/compaction_group.hh:
```diff
@@ -62,7 +68,11 @@ private:
+    compaction_group(table& t, idx_type idx, dht::token_range token_range);
+
+    idx_type idx() const noexcept {
+        return _idx;
```
perhaps idx_type index()? I know we're well familiar with what idx stands for, but full name leaves no doubt. e.g. google tells me it stands for "Internet Data Exchange" :-)

raphaelsc replied:
This is some bikeshedding from me though, feel free to ignore it.

bhalevy replied:
changing that in next version to present a fully fledged `struct group_id` that contains both idx and total.
I think it makes more sense, once exposed also for table state

raphaelsc commented:
Given that you introduced comapction group :: index in this series, I somehow expected it to integrate the index more in compaction layer, but sounds like you just wanted to get it started, which is good.

Series look good to me in general, just had a few comments.

bhalevy commented:
In v2:
- kept skipping of offstrategy compaction in compaction_manager layer
- added struct group_id (in a new compaction/group_id.hh header)
- changed all relevant printouts in compaction_manager.cc to use a formatter from table_state that prints both the ks.cf names and the compaction group_id in the form of "idx/total"

raphaelsc commented on compaction/group_id.hh:
```diff
@@ -0,0 +1,59 @@
+
+    group_id& operator++() noexcept {
+        ++idx;
+        return *this;
```
an improvement for later is to check if we're going out of valid range and throw std::out_of_range if so.

raphaelsc commented on compaction/table_state.hh:
```diff
@@ -54,5 +54,17 @@ public:
+        return format_to(ctx.out(), "{}.{} compaction_group={}", s->ks_name(), s->cf_name(), t.gid());
+    }
+};
 
```
very good

avikivity commented on compaction/group_id.hh:
```diff
@@ -0,0 +1,59 @@
+
+    group_id() noexcept : idx(0), total(0) {}
+    group_id(unsigned idx, unsigned total) noexcept : idx(idx), total(total) {}
+    explicit group_id(unsigned total) noexcept : idx(0), total(total) {}
```
Why this constructor? Seems wrong.

avikivity commented on replica/table.cc:
```diff
@@ -526,8 +527,9 @@ std::vector<std::unique_ptr<compaction_group>> table::make_compaction_groups() {
     std::vector<std::unique_ptr<compaction_group>> ret;
     auto&& ranges = dht::split_token_range_msb(_x_log2_compaction_groups);
     tlogger.debug("Created {} compaction groups for {}.{}", ranges.size(), _schema->ks_name(), _schema->cf_name());
+    auto gid = compaction::group_id(ranges.size());
```
Okay, I see why you added that constructor. But it's not intuitive. Better to initialize it from 0, size.

avikivity replied:
Or even, instead of a funky type just have a string there. Then we can later put the table name + tablet id, or whatever. Only this emplace_back needs to worry about it.

bhalevy replied:
okay, sounds good

bhalevy commented:
v3 ([9768046](https://github.com/scylladb/scylladb/commit/9768046d7cb69a4391bdb98e0ddbf8a9bc1e84c9)):
- dropped struct group_id
  - use a std::string group id member instead.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/838/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/838/

https://github.com/scylladb/scylladb/issues/13645

will retry

avikivity commented:
did not work

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/929/

bhalevy commented:
@scylladb/scylla-maint please merge
